File: week13/13/CarPooling.cpp
class Solution {
public:
    bool carPooling(vector<vector<int>>& trips, int capacity) {
        
        vector<int> num(1001, 0);

        for(auto i=0; i<trips.size(); i++){
            for(auto j=trips[i][1]; j<trips[i][2]; j++){
                num[j]+=trips[i][0];
            }
        }

        for(auto i=0; i<=1000; i++){
            if(num[i]>capacity){
                return false;
            }
        }
        return true;
    }
};


//map을 사용해서 아래와 같이 공간 효율적으로 구현할 수도 있음

//#include <vector>
// #include <map>
// using namespace std;

// class Solution {
// public:
//     bool carPooling(vector<vector<int>>& trips, int capacity) {
//         map<int, int> changes; // 변화 기록 (위치 기반 정렬 자동 처리)

//         // 승객 태우고 내리기 기록
//         for (auto& trip : trips) {
//             int numPassengers = trip[0];
//             int from = trip[1];
//             int to = trip[2];

//             changes[from] += numPassengers; // from 위치에서 승객 태움
//             changes[to] -= numPassengers;  // to 위치에서 승객 내림
//         }

//         int currentPassengers = 0;

//         // 누적된 승객 수 확인
//         for (auto& change : changes) {
//             currentPassengers += change.second; // 현재 위치의 승객 변화 적용
//             if (currentPassengers > capacity) {
//                 return false; // 용량 초과
//             }
//         }

//         return true; // 모든 승객을 태울 수 있음
//     }
// };

File: week13/29/carPooling.py
"""
<Car Pooling>

There is a car with capacity empty seats. 
The vehicle only drives east (i.e., it cannot turn around and drive west).

You are given the integer capacity and an array trips 
where trips[i] = [numPassengers_i, from_i, to_i] indicates that the ith trip has numPassengers_i passengers 
and the locations to pick them up and drop them off are from_i and to_i respectively. 
The locations are given as the number of kilometers due east from the car's initial location.

Return true if it is possible to pick up and drop off all passengers for all the given trips, or false otherwise.
"""

def carPooling(trips,capacity):
    """
    Record passenger changes at each location (from_i에서 추가, to_i에서 빼기)
    Check if the cumulative count exceeds the car's capacity at any point
    """
    passenger_changes = {}

    for num_passengers, start, end in trips:
        passenger_changes[start] = passenger_changes.get(start,0) + num_passengers
        passenger_changes[end] = passenger_changes.get(end,0) - num_passengers
        
    current_passengers = 0
    for location in sorted(passenger_changes.keys()):   # location 순서대로 (가까운 곳 -> 먼 곳) 누적값 계산
        current_passengers += passenger_changes[location]
        if current_passengers > capacity:
            return False

    return True


print(carPooling([[2,1,5],[3,3,7]],4))  # false
print(carPooling([[2,1,5],[3,3,7]],5))  # true